runtime/script_loader.cpp

```diff
@@ -71,20 +71,56 @@ static const char* resolve_extension(const char* resolved_path) {
 
 static const char* resolve_path(const char* path, const char* base, size_t base_len) {
   MOZ_ASSERT(base);
-  if (path[0] == '/') {
-    return resolve_extension(strdup(path));
-  }
-  if (path[0] == '.' && path[1] == '/') {
-    path = path + 2;
-  }
   while (base_len > 0 && base[base_len - 1] != '/') {
     base_len--;
   }
-  size_t len = base_len + strlen(path) + 1;
-  char* resolved_path = new char[len];
-  strncpy(resolved_path, base, base_len);
-  strncpy(resolved_path + base_len, path, len - base_len);
-  MOZ_ASSERT(strlen(resolved_path) == len - 1);
+  size_t path_len = strlen(path);
+
+  // create the maximum buffer size as a working buffer
+  char* resolved_path = new char[base_len + path_len + 1];
+
+  // copy the base in if used
+  size_t resolved_len = base_len;
+  if (path[0] == '/') {
+    resolved_len = 0;
+  } else {
+    strncpy(resolved_path, base, base_len);
+  }
+
+  // then iterate by segment of the path
```

Comment:
Can you expand this comment to explain what the loop actually does? It's not trivial to follow, and seems pretty relevant to getting the right behavior here.

Reply:
Sure, added some more comments.